nncf/experimental/common/pruning/operations.py

```diff
@@ -0,0 +1,637 @@
+                    if input_channels in list_output_channels:
+                        index = list_output_channels.index(input_channels)
+                        shifted_dim = inp_map[dim][index]
+                        output_mask.dim_groups_map[shifted_dim] = groups
+                    else:
+                        group = groups[0]
+                        if group.has_children():
+                            raise NotImplementedError('Splitting BlockGroup with childs isn\'t implemented yet')
+                        child_groups = cls._split_group(group, input_channels, list_output_channels)
+                        for child_group, in_dim in zip(child_groups, inp_map[dim]):
+                            output_mask.dim_groups_map[in_dim] = [child_group]
+            return output_mask
+
+        if mode == ReshapeMode.SHRINK:
+            grouping = defaultdict(list)
+            for inp_idx, groups in mask.dim_groups_map.items():
+                grouping[inp_map[inp_idx]].extend(groups)
+            output_mask.dim_groups_map = dict(grouping)
+            return output_mask
+
+
+class TransposePruningOp(BasePruningOp):
+    @classmethod
+    def mask_propagation(cls, node: NNCFNode, graph: NNCFGraph,
+                         tensor_processor: Type[NNCFPruningBaseTensorProcessor]) -> None:
+        input_masks = get_input_masks(node, graph)
+        assert len(input_masks) == 1
+        input_mask = input_masks[0]
+        if input_mask is None:
+            node.data['output_mask'] = None
+            return
+
+        if isinstance(node.layer_attributes, TransposeLayerAttributes):
+            input_tensors_shapes = [x.tensor_shape for x in graph.get_input_edges(node)]
+            assert len(input_tensors_shapes) == 1
+            new_order = list(range(len(input_tensors_shapes[0])))
+            dim0 = node.layer_attributes.dim0
+            dim1 = node.layer_attributes.dim1
+            new_order[dim1], new_order[dim0] = new_order[dim0], new_order[dim1]
+        elif isinstance(node.layer_attributes, PermuteLayerAttributes):
+            new_order = node.layer_attributes.permutation
+
+        idx_map = [(old_idx, new_idx) for new_idx, old_idx in enumerate(new_order) if
+                   old_idx in input_mask.dim_groups_map]
+        output_mask = PropagationMask(
+            dim_groups_map={
+                new_idx: input_mask.dim_groups_map[old_idx] for old_idx, new_idx in idx_map
+            }
+        )
+
+        node.data['output_mask'] = output_mask
+
+
+class StopMaskForwardPruningOp(BasePruningOp):
+    @classmethod
+    def accept_pruned_input(cls, node: NNCFNode) -> bool:
+        return False
+
+    @classmethod
+    def mask_propagation(cls, node: NNCFNode, graph: NNCFGraph,
+                         tensor_processor: Type[NNCFPruningBaseTensorProcessor]) -> None:
+        input_masks = get_input_masks(node, graph)
+        for m in input_masks:
+            if m:
+                m.invalidate_groups()
```

Comment:
Can we cover this repeating code by a function with a callable? Something like `for_each_not_none`? I believe it would be usefull